@s3/store: add backpressure + download incomplete part first

fenos opened the pull request:
This PR improves the handling of incomplete parts and adds backpressure by limiting the number of concurrent uploads/files on the disk that are being created.

## Incomplete Parts

The incomplete part logic was somewhat not always working as intended.

An issue I came across with the old implementation is that when i upload a file using chunked upload from the client, and the chunk is bigger than the `preferred` size on the server, at times the offset returned wasn't calculated properly resulting in the subsequent request to return 409 (Offset-Conflict) very much related to  #501

This PR fixes this issue, by always downloading the incomplete part first and prepending it to the readstream.
By using this approach the offset is now calculated properly and also fixes #501 (see test) 


## Backpressure

Added support for backpressure. The way it works is by using a semaphore on the StreamSplitter.
Only X number of chunks on disk can be created at any one time (default 30) and the semaphore is released upon finishing the part upload, meaning that the backpressure is handled by how fast S3 can upload parts.

In the scenario when the limit of concurrent chunks has been reached, the default node mechanism of backpressure will kick in. This means that when the highwater mark on the readstream buffer is full,  the stream will be paused until the flow of data continues and the semaphore has capacity.

This aligns with the issue #505

fenos commented on packages/server/src/handlers/BaseHandler.ts:
```diff
@@ -151,9 +151,11 @@ export class BaseHandler extends EventEmitter {
-          proxy.destroy(err)
+          // we end the stream gracefully here so that we can upload the remaining bytes to the store
+          // as an incompletePart
+          proxy.end()
```
A very small but nice change here.
When an upload request is canceled, Instead of calling `proxy.destroy()` we can call `proxy.end()` to gracefully close the passthrough stream, so that we can flush the bytes of a chunk as an incomplete part and upload it to S3.

This will result in starting exactly where we left off as opposed to starting from the last successful chunk uploaded.

Acconut replied:
Nice!

Murderlon commented on packages/s3-store/index.ts:
```diff
@@ -25,6 +28,7 @@ type Options = {
   // but may increase it to not exceed the S3 10K parts limit.
   partSize?: number
   useTags?: boolean
+  maxConcurrentPartUploads?: number
```
Do we need to expose this as an option? I think this falls under the category of we as upload experts should determine what is best for users for this low-level option.

fenos replied:
It all depends on the upload bandwidth of where the server is running, there is no magic value to be fair.
At the moment we are simply limiting how many parallel chunks and uploads can be processed at any point in time. 

For example, if you are running the server in a local network with a local minio, this value could be much higher since the upload throughput is higher.

fenos replied:
similarly, you run the uploads within the AWS platform (ECS/ECS, etc..) against S3, the upload would be faster than if you upload it from a server in a different provider. Meaning the concurrentPartUploads could increased.

My default value of max 30 concurrent chunks with a preferred size of 8MB means that the upload bandwidth will be around 240MB/s - the file system will match that with backpressure. 

Without backpressure, the file system could write parts at a speed of 1GB/s but upload at much slower rate

Murderlon replied:
That would be a bit too naive to not expose it. Perhaps we want to write some docs about the considerations, as I do think tus-node-server is also popular under front-end/full-stack devs with less awareness of this. 

how about something like this in the s3-store readme:

### Example: using `partSize` and `maxConcurrentParts` efficiently

These settings depend on the machine's processing power, memory, network bandwidth, and disk I/O capabilities.
Below are some recommendations to take into account but exact settings depend on your setup.

- **Uploading within the AWS platform to S3**. Given the high-speed environment and assuming robust server resources both values can be set higher than the default.
- **Uploading from outside the AWS platform to S3**.  These are variable-speed environments with more network fluctuations. The default `partSize` of 8 MiB and default `maxConcurrentParts` of 30 could work well for this (or lower on a resource-limited machine).
- **Using Local MinIO** (Local Network or On-Premises). Typically high-speed and low-latency but this should be balanced against the capabilities of your local server infrastructure. Values similar to or higher than those used within AWS might be suitable.

In general, the tradeoff is reliability vs. speed. A bigger `partSize` requires less roundtrips but is more prone to failure in less stable connections. Higher concurrency (`maxConcurrentParts`) increases CPU and memory usage.

Murderlon commented on packages/s3-store/index.ts:
```diff
@@ -741,4 +809,26 @@ export class S3Store extends DataStore {
+    const fileName = template + crypto.randomBytes(10).toString('base64url').slice(0, 10)
+    const filePath = path.join(os.tmpdir(), fileName)
+
+    const fileExists = await new Promise<boolean>((resolve, reject) => {
+      fs.lstat(filePath, (error) => {
+        if (!error) {
+          return resolve(true)
+        }
+        if (error.code === 'ENOENT') {
+          return resolve(false)
+        }
+        reject(error)
+      })
+    })
```
Would be nice to start to transition to `fs/promises` where possible, especially new additions.

fenos replied:
Yes, I just found it less verbose using the callback instead of `try { } catch {} ` for this instance.
But happy to migrate to fs/promises if preferred

fenos replied:
Done

Murderlon reviewed (commented):
A great addition!

Might need to go over it one more time but some initial feedback here.

fenos commented on packages/server/src/models/StreamSplitter.ts:
```diff
@@ -117,13 +125,34 @@ export class StreamSplitter extends stream.Writable {
   }
 
+  async emitEvent<T>(name: string, payload: T) {
+    if (this.asyncEvents) {
```
a bit less canonical but working with async events from the event emitter is also not that great 😄

fenos replied:
Oh! 
I just learned a neat wait to have async compatible events!

instead of calling `emit('name', payload)`

we implement our own `emitAsync` function:

```ts
  async emitEvent<T>(name: string, payload: T) {
    const listeners = this.listeners(name)
    for await (const listener of listeners) {
      await listener(payload)
    }
  }
```

and voila'

fenos commented:
@Murderlon I've updated the Readme and added a detailed documentation on the `maxConcurrentPartUploads` flag.

I've also increased the value to 60 instead of 30 (~480MB), as this aligns better with tusd default.
Tusd uses 50MB `partSize` and 10 `maxConcurrentPartUploads` totaling at 500MB.

Let me know if there is anything else, this change will be awesome to have in

Murderlon commented on packages/s3-store/README.md:
```diff
@@ -91,6 +91,22 @@ you need to provide a cache implementation that is shared between all instances
 
 See the exported [KV stores][kvstores] from `@tus/server` for more information.
 
+### `options.maxConcurrentPartUploads`
```
```suggestion
#### `options.maxConcurrentPartUploads`
```

Murderlon commented on packages/s3-store/README.md:
```diff
@@ -91,6 +91,22 @@ you need to provide a cache implementation that is shared between all instances
+The default value is 60. This default is chosen in conjunction with the typical partSize of 8MiB, aiming for an effective transfer rate of approximately 480MiB/s.
+
+**Considerations:**
+The ideal value for maxConcurrentPartUploads varies based on your `partSize` and the upload bandwidth to your S3 bucket. A larger partSize means less overall upload bandwidth available for other concurrent uploads.
+
+- **Lowering the Value**: Reducing maxConcurrentPartUploads decreases the number of simultaneous upload requests to S3. This can be beneficial for conserving memory, CPU, and disk I/O resources, especially in environments with limited system resources or where the upload speed it low or the part size is large.
+
+
+- **Increasing the Value**: A higher value potentially enhances the data transfer rate to the server, but at the cost of increased resource usage (memory, CPU, and disk I/O). This can be advantageous when the goal is to maximize throughput, and sufficient system resources are available.
+
+
+- **Bandwidth Considerations**: It's important to note that if your upload bandwidth to S3 is a limiting factor, increasing maxConcurrentPartUploads won’t lead to higher throughput. Instead, it will result in additional resource consumption without proportional gains in transfer speed.
```
```suggestion
The ideal value for `maxConcurrentPartUploads` varies based on your `partSize` and the upload bandwidth to your S3 bucket. A larger partSize means less overall upload bandwidth available for other concurrent uploads.

- **Lowering the Value**: Reducing `maxConcurrentPartUploads` decreases the number of simultaneous upload requests to S3. This can be beneficial for conserving memory, CPU, and disk I/O resources, especially in environments with limited system resources or where the upload speed it low or the part size is large.


- **Increasing the Value**: A higher value potentially enhances the data transfer rate to the server, but at the cost of increased resource usage (memory, CPU, and disk I/O). This can be advantageous when the goal is to maximize throughput, and sufficient system resources are available.


- **Bandwidth Considerations**: It's important to note that if your upload bandwidth to S3 is a limiting factor, increasing `maxConcurrentPartUploads` won’t lead to higher throughput. Instead, it will result in additional resource consumption without proportional gains in transfer speed.
```

Murderlon reviewed (approved):
LGTM 👌 

I think I would like a quick review from @Acconut on this before merging.

Acconut commented on packages/s3-store/README.md:
```diff
@@ -91,6 +91,22 @@ you need to provide a cache implementation that is shared between all instances
 
+#### `options.maxConcurrentPartUploads`
+
+This setting determines the maximum number of simultaneous part uploads to an S3 storage service.
```
It might be helpful to state in this document what a part actually is. Maybe a brief explanation saying that this module uses S3 multipart uploads, which splits a file into multiple equally-sized parts (independent of the PATCH requests). With this context, it's easier to understand what concurrent part uploads actually means.

Acconut commented on packages/s3-store/README.md:
```diff
@@ -91,6 +91,22 @@ you need to provide a cache implementation that is shared between all instances
+#### `options.maxConcurrentPartUploads`
+
+This setting determines the maximum number of simultaneous part uploads to an S3 storage service.
+The default value is 60. This default is chosen in conjunction with the typical partSize of 8MiB, aiming for an effective transfer rate of approximately 480MiB/s.
```
The calculation seems assume that one part is uploaded in 1s. Might be worth mentioning. Also transfer speeds are commonly given in Mbit/s not MiB/s. That would be 3.84 Gbit/s, which seems very high. I am not sure if the transfer speeds to S3 reach that level.

fenos replied:
Traffic between Amazon EC2 and Amazon S3 uses up to 100 Gbps of bandwidth to Amazon Virtual Private Cloud (Amazon VPC) endpoints and public IP addresses in the same AWS Region.

Acconut commented on packages/s3-store/index.ts:
```diff
@@ -42,6 +46,45 @@ function calcOffsetFromParts(parts?: Array<AWS.Part>) {
   return parts && parts.length > 0 ? parts.reduce((a, b) => a + b.Size, 0) : 0
 }
 
+function concatenateStreams(streams: Readable[]): Readable {
```
Is it possible to use an existing module for that, e.g. https://www.npmjs.com/package/multistream? Then we would not have to maintain this logic.

fenos replied:
Ah nice! sure happy to use `multistream` in this case

Murderlon replied:
If this single function does the job, I personally don't see the reason to add more dependencies, unless there are other benefits.

Acconut commented on packages/s3-store/index.ts:
```diff
@@ -82,6 +125,7 @@ export class S3Store extends DataStore {
   private preferredPartSize: number
   private expirationPeriodInMilliseconds = 0
   private useTags = true
+  private semaphore: Semaphore
```
```suggestion
  private partUploadSemaphore: Semaphore
```

A good name helps understand what the semaphore actually controls.

Acconut commented on packages/s3-store/index.ts:
```diff
@@ -233,6 +278,61 @@ export class S3Store extends DataStore {
+    try {
+      let incompletePartSize = 0
+
+      const byteCounterTransform = new stream.Transform({
```
Using a transformer to obtain the incomplete part's size appear a bit complex to me. Two alternatives are either using the length that is returned in the `GetObject` call to S3 (AFAIK there should be a length in there) or reading the file size after the part has been written to disk. Then we can remove this additional stream.

Acconut replied:
I just saw that this code was copied from somewhere else. Feel free to ignore it then :)

Acconut commented on packages/s3-store/index.ts:
```diff
@@ -484,9 +536,11 @@ export class S3Store extends DataStore {
       optimalPartSize = this.preferredPartSize
       // The upload is too big for the preferred size.
-      // We devide the size with the max amount of parts and round it up.
+      // We divide the size with the max amount of parts and round it up.
```
There is no rounding up here happening. The comment likely needs to be updated.

Acconut commented on packages/s3-store/index.ts:
```diff
@@ -484,9 +536,11 @@ export class S3Store extends DataStore {
+      optimalPartSize = size / this.maxMultipartParts
     } else {
-      optimalPartSize = Math.ceil(size / this.maxMultipartParts)
+      optimalPartSize = size / this.maxMultipartParts + 1
```
Why was the `Math.ceil` removed? If we reach this conditional block, we know that `size / this.maxMultipartParts` is not an integer, so rounding it would be sense to get rid of the fractional part.

fenos replied:
I based this change on this:
https://github.com/tus/tusd/blob/main/pkg/s3store/s3store.go#L1176

There is no rounding as far as i can see

Murderlon replied:
Go automatically does rounding :)

Acconut replied:
To add some more explanation to @Murderlon's comment: The code from tusd uses integers. Integer division is different than floating point division in that the fractional is ignored. For example:

```
14 / 5 == 2 // integer division
14.0 / 5 == 2.8 // float divison
```

That's why you want `Math.floor` when trying to replicate the Go logic, or `Math.ceil` when you want to replicate the `integer division + 1` from tusd.

fenos replied:
Ah yes, very good point! I totally missed that the data types were ints and not float. Will bring the rounding back :)

Acconut commented on packages/s3-store/index.ts:
```diff
@@ -233,6 +278,61 @@ export class S3Store extends DataStore {
+        fs.createWriteStream(filePath)
+      )
+
+      const createReadStream = (options: {cleanUpOnEnd: boolean}) => {
```
`cleanUpOnEnd` seems to always be true. Do we need this option then?

fenos replied:
Yhea, this option is more for awareness, that at the end of the stream, the file is removed.

It could hide the detail in which the file is being removed to the next developer looking at the code if we simply call: `createReadStream()` without the option.

Happy to remove it, but it seems clearer with it

Acconut commented on packages/s3-store/index.ts:
```diff
@@ -741,4 +799,29 @@ export class S3Store extends DataStore {
     return deleted
   }
+
+  private async uniqueTmpFileName(template: string): Promise<string> {
```
Is there also a public package that we could reuse here?

fenos replied:
i think this is such a basic functionality that a package might be not necessary, wdyt?

Acconut commented on packages/server/src/models/Semaphore.ts:
```diff
@@ -0,0 +1,63 @@
+  promise?: Promise<Permit>
+}
+
+export class Semaphore {
```
Nice implementation, but I am wondering again whether we can reuse existing module for this.

fenos replied:
Yea, I thought that minimizing dependencies for these utility classes could be ok

Acconut commented on packages/server/src/models/StreamSplitter.ts:
```diff
@@ -106,7 +108,7 @@ export class StreamSplitter extends stream.Writable {
     await this.fileHandle.close()
 
-    this.emit('chunkFinished', {
+    await this.emitEvent('chunkFinished', {
```
Can you explain why this change was necessary?

fenos replied:
to keep the events interface consistent

Acconut commented on packages/server/src/models/StreamSplitter.ts:
```diff
@@ -117,13 +119,23 @@ export class StreamSplitter extends stream.Writable {
     this.part += 1
   }
 
+  async emitEvent<T>(name: string, payload: T) {
```
This function should be named differently because it doesn't emit events in an asynchronous fashion like we are used from EventEmitter. Instead, it sequentially invokes callbacks, which is different.

fenos replied:
the standard EventEmitter emits events **synchronously** this function does the same thing but it allows using async/await to have a similar behavior to the standard EventEmitter

fenos replied:
In essence the standard EventEmitter loops sequentially all events as my implementation and the handler blocks in a synchronous fashion, however, if you do async operations it won't await by default

Acconut replied:
You are right, thanks for correct me! Then the name is fine.

Acconut reviewed (changes requested):
Amazing work, @fenos!

Acconut commented:
Regarding replacing internal code with dependencies, I would wait for a comment from @Murderlon. Maybe his position is different than mine :)

Murderlon commented:
> Regarding replacing internal code with dependencies, I would wait for a comment from @Murderlon. Maybe his position is different than mine :)

I'd personally would avoid adding a dependency if it's not much code, we likely rarely/never have to change it, and we don't foresee complex side-effects. So I don't think we need a package to concat streams or unique tmp files. For the semaphore I don't have a strong opinion, in this case it's a well tested copy paste from Shopify so from a trust perspective I'd say it's equal to a package. But if there is a similar well tested package with a small footprint I would be okay with that too. I'm okay with either.

Acconut commented:
Alright, if you prefer less dependencies, we can go with that route as well. That's probably a topic where personal preferences differ.

Murderlon commented:
I just saw that [`@shopify/semaphore`](https://www.npmjs.com/package/@shopify/semaphore) is available as a package. In that case I agree with @Acconut that a dependency is better.

Murderlon commented on packages/s3-store/index.ts:
```diff
@@ -25,6 +28,7 @@ type Options = {
   // but may increase it to not exceed the S3 10K parts limit.
   partSize?: number
   useTags?: boolean
+  maxConcurrentPartUploads?: number | Semaphore
```
It's not documented that this can also be a `Semaphore`. Any reason against keeping this simple and only doing a `number`, which we always pass to `Semaphore`?

fenos replied:
I thought that in case a server runs multiple instances of TusServer they might want to share a single Semaphore (which I think in the future we should also allow passing an S3Client too)

Will document now

Murderlon replied:
Is that something Supabase already needs? I'm also okay with keeping it simple for now until the need arrises, then we would add a new option which would also be less odd than abusing `maxConcurrentPartUploads` a bit.

fenos replied:
We don't need it right now, but probably in the near future.
Happy to revisit this later

fenos replied:
Reverted this

Murderlon reviewed (commented):
Changes look good 👍 One Q

fenos commented:
@Murderlon all ready

Murderlon reviewed (approved):
Thanks!
